File: ts/src/data-structures/tree/search-bst/search.test.ts
import {test, expect} from "bun:test";

import {searchBST, searchBST2, searchBSTV3} from "./search";
import {TreeNode} from "../TreeNode";

function nodesToArray(root: TreeNode | null): number[] {
  if (root === null) return [];
  return [root.val, ...nodesToArray(root.left), ...nodesToArray(root.right)];
}

test("searchBST", () => {
  let root = new TreeNode(4);
  root.left = new TreeNode(2);
  root.right = new TreeNode(7);
  root.left.left = new TreeNode(1);
  root.left.right = new TreeNode(3);

  expect(nodesToArray(searchBST(root, 2))).toEqual([2, 1, 3]);
  expect(nodesToArray(searchBST2(root, 2))).toEqual([2, 1, 3]);
  expect(nodesToArray(searchBSTV3(root, 2))).toEqual([2, 1, 3]);
});

File: src/algorithms/mixed/search-bst/search_test.go
package mixed

import (
	"testing"
)

func TestSearchBST(t *testing.T) {
	root := &TreeNode{Val: 4}
	root.Left = &TreeNode{Val: 2}
	root.Right = &TreeNode{Val: 7}
	root.Left.Left = &TreeNode{Val: 1}
	root.Left.Right = &TreeNode{Val: 3}

	tests := []struct {
		root *TreeNode
		val  int
		want *TreeNode
	}{
		{root, 2, root.Left},
		{root, 5, nil},
	}

	for _, test := range tests {
		got := searchBST(test.root, test.val)
		if got != test.want {
			t.Errorf("searchBST(%v, %d) = %v, want %v", test.root, test.val, got, test.want)
		}
	}
}

func TestSearchBST2(t *testing.T) {
	root := &TreeNode{Val: 4}
	root.Left = &TreeNode{Val: 2}
	root.Right = &TreeNode{Val: 7}
	root.Left.Left = &TreeNode{Val: 1}
	root.Left.Right = &TreeNode{Val: 3}

	tests := []struct {
		root *TreeNode
		val  int
		want *TreeNode
	}{
		{root, 2, root.Left},
		{root, 5, nil},
	}

	for _, test := range tests {
		got := searchBST(test.root, test.val)
		if got != test.want {
			t.Errorf("searchBST(%v, %d) = %v, want %v", test.root, test.val, got, test.want)
		}
	}
}
